### DISCUSSION TITLE
Using assign on an entity with composite key?

### DISCUSSION THREAD

**Discussion by nurdism** [mikro-orm/mikro-orm#2759] [Q&A]
Is there a way to use `assign()` with an entity with composite key? When I attempt to do this it throws `ValidationError: Composite key required for entity`. This will throw even if I pass both required PKs. 

Example:
```ts
@Entity()
export class TemplateCategoryRelation extends BaseEntity<TemplateCategoryRelation, 'template' | 'templateCategory'> {
  [PrimaryKeyType]?: [string, string]

  @ManyToOne((type) => Template, { inversedBy: (e) => e.templateCategoryRelationCollection, onDelete: 'CASCADE', primary: true })
  template: Template

  @ManyToOne((type) => TemplateCategory, { inversedBy: (e) => e.templateCategoryRelationCollection, onDelete: 'CASCADE', primary: true })
  templateCategory: TemplateCategory
}

@Entity()
export class Template extends BaseEntity<Template, 'id'> {
  @PrimaryKey({ type: BigIntType })
  id: string

  @Property({ type: StringType, length: 255, nullable: true })
  title: string | null

  @OneToMany((type) => TemplateCategoryRelation, (e) => e.template)
  templateCategoryRelationCollection = new Collection<TemplateCategoryRelation>(this)
}

@Entity()
export class TemplateCategory extends BaseEntity<TemplateCategory, 'id'> {
  @PrimaryKey({ type: BigIntType })
  id: string

  @Property({ type: StringType, length: 255, nullable: true })
  title: string | null

  @OneToMany((type) => TemplateCategoryRelation, (e) => e.templateCategory)
  templateCategoryRelationCollection = new Collection<TemplateCategoryRelation>(this)
}

const categories: TemplateCategory[] = []; // find or create
const template = new Template()

template.assign(
  {
    title: 'foo'
    templateCategoryRelationCollection: categories.map((templateCategory) => ({ template, templateCategory })),
  },
  { em: req.em },
)

//ValidationError: Composite key required for entity TemplateCategoryRelation.
//    at Function.invalidCompositeIdentifier (/workspace/node_modules/@mikro-orm/core/errors.js:77:16)
//    at SqlEntityManager.getReference (/workspace/node_modules/@mikro-orm/core/EntityManager.js:462:48)
//    at /workspace/node_modules/@mikro-orm/core/entity/EntityAssigner.js:127:36
//    at Array.map (<anonymous>)
//    at Function.assignCollection (/workspace/node_modules/@mikro-orm/core/entity/EntityAssigner.js:123:29)
//    at /workspace/node_modules/@mikro-orm/core/entity/EntityAssigner.js:33:39
//    at Array.forEach (<anonymous>)
//    at Function.assign (/workspace/node_modules/@mikro-orm/core/entity/EntityAssigner.js:24:27)
//    at Template.assign (/workspace/node_modules/@mikro-orm/core/entity/BaseEntity.js:29:48)```

**Comment by B4nan** [comment]
>  When I attempt to do this it throws

So when you do _what_? Share some code.

**Reply by nurdism** [reply]
You're right, updated my post

**Reply by B4nan** [reply]
What you are trying to do, set the FK or do deep entity graph update? Is this v4 or v5?

To set a FK, you should be passing ordered tuple of the composite PKs

```ts
    templateCategoryRelationCollection: categories.map((templateCategory) => ([template, templateCategory])),
```

Maybe this is too strict (or malfunctioning) validation.

**Reply by nurdism** [reply]
I'm on v5, when attempting to pass an array TS throws
```
(property) templateCategoryRelationCollection?: EntityDataItem<Collection<TemplateCategoryRelation, unknown>> | undefined
Type '[Template, TemplateCategory][]' is not assignable to type 'EntityDataItem<Collection<TemplateCategoryRelation, unknown>> | undefined'.
  Type '[Template, TemplateCategory][]' is not assignable to type 'TemplateCategoryRelation[]'.
    Type '[Template, TemplateCategory]' is not assignable to type 'TemplateCategoryRelation'.ts(2322)
```

I tried it with a `//@ts-ignore` to see if it runs. It throws `Invalid collection values provided for 'Template.templateCategoryRelationCollection' in Template.assign()`

**Reply by nurdism** [reply]
I do have a workaround I just 
```ts
for (const templateCategory of categories) {
  template.templateCategoryRelationCollection.add(TemplateCategoryRelation.create({ template, templateCategory }, { em: req.em }))
}
```

But it would be a "nice to have", to be able to assign directly to the template

**Reply by B4nan** [reply]
Oh right, so you are trying to assing composite PK to a collection, will take a look when I have some spare time, but I think this is not supported right now.

> I do have a workaround I just

Note that this is not really equal to using assign helper with an array of PKs - that would replace the collection items, not just add new ones. Also it would just set the PKs - they would be required to represent existing entity. Your workaround is actually creating new target entities, that woulnd't be done with assign at all.

**Reply by nurdism** [reply]
I may just be using `assign()` wrong in that case, what I am trying to do is create new `TemplateCategoryRelation[]` 

I have a array of `TemplateCategory` and I want to create the `TemplateCategoryRelation` by passing a `TemplateCategory` to the assign function. I assumed this would create a `TemplateCategoryRelation` and add it to the `templateCategoryRelationCollection` field in the template. 

This does work with entities with a single pk, but it might not be working the way I think it is.

**Reply by B4nan** [reply]
Yeah I think this is not supported for composite keys, and what you were doing originally would be the right way. The suggestion with tuples is only if you want to set the FK to existing entity, not if you want to create deep entity graph.

I will debug this more and see if we can make it work, maybe it is really just too strict validation and things would work if we remove it.

Note that assign would still _replace_ the items, not just _add_ new ones.

And lastly, you should _not_ use `assign` on new entities, this was never the use case for it. Use should `em.create()` here.

```ts
const template = req.em.create(Template, {
  title: 'foo'
  templateCategoryRelationCollection: categories, // the template reference will get propagated automatically
})
```

**Reply by nurdism** [reply]
I've tried to use `req.em.create(...)` but the `RequiredEntityData<T>` is too strict. In my entities I define read only "get" properties and the `RequiredEntityData<T>` makes them required, it also makes fields like `createdAt` that have a default, required. 

I use the `new Entity().assign(data, { em })` to avoid the type errors like that. There are some cases I don't have all the data to create the entity but want to create the reference.  I'm not sure if this is an un indented side affect, but this is how I've been creating my entities

```ts
@Entity()
export class Organization extends BaseEntity<Organization, 'id'> {
  static create(data: EntityData<Organization>, options: AssignOptions & { em: EntityManager }): Organization {
    return new Organization().assign(data, options)
  }

  @PrimaryKey({ type: SnowFlakeType, defaultRaw: 'snowflake()', autoincrement: false })
  id: string

  @Field({ description: '' })
  @Property({ type: StringType, length: 64 })
  title: string

  @OneToMany((type) => OrganizationMergetag, (e) => e.organization)
  organizationMergetagCollection = new Collection<OrganizationMergetag>(this)
}

@Entity()
export class OrganizationMergetag extends BaseEntity<OrganizationMergetag, 'id'> {
  @PrimaryKey({ type: SnowFlakeType, defaultRaw: 'snowflake()', autoincrement: false })
  id: string

  @ManyToOne((type) => Organization, { inversedBy: (e) => e.organizationMergetagCollection, onDelete: 'CASCADE' })
  organization: Organization

  @Property({ type: StringType, length: 255, nullable: true })
  title: string | null = null
}

const organization = Organization.create({
  title,
  organizationMergetagCollection: [
    {
      title: 'First Name',
    },
    {
      title: 'Last Name',
    },
  ],
}, { em })

await em.persistAndFlush(organization)
``` 

Will persist the Organization and each of the OrganizationMergetag, and in most cases this works perfectly fine.

**Reply by B4nan** [reply]
>  In my entities I define read only "get" properties and the RequiredEntityData<T> makes them required, it also makes fields like createdAt that have a default, required.

You need to use `OptionalProps` for those cases. Anything that is not defined as optional is required unless you add it there, we can't infer any information from decorators or see if there is a property initializer or not.

**Reply by nurdism** [reply]
Oh perfect, is there any documentation on how to use OptionalProps? I searched the docs and only found one mention of it in the upgrade guide from V4 to V5. Is it just used like `[OptionalProps]?: ["PropOne", "PropTwo"]`?

**Reply by B4nan** [reply]
I will need to add a new section about type safety, I know its missing (and there are more symbols than just `OptionalProps`, e.g. the ones for composite PKs, repository type, I guess this will go to the defining entities section and maybe a separate section just about the type safety (e.g. how `Loaded` type and populate hints work).

> Is it just used like [OptionalProps]?: ["PropOne", "PropTwo"]?

It is a union of keys, not an array, so `[OptionalProps]?: "PropOne" | "PropTwo"`.

**Reply by nurdism** [reply]
Awesomeness! Thank you very much for your help and for your great work you've done with this library. One of the best ORMs I've had the pleasure of using.
